### ISSUE TITLE
WidgetTester's "end of test verifications" cause integration tests to always fail - still an issue in the emulator

### ISSUE THREAD

**Issue by nc-jeni** [flutter/flutter#142713]
### Steps to reproduce

_The error occurs only in the emulator, not on a real device!_

1. Clone [the Patrol repository](https://github.com/leancodepl/patrol) and cd into packages/patrol/example
2. Start emulator `Resizable Experimental API 34 (mobile)`
3. flutter run to generate gradlew and to make sure that the example app builds and runs just fine
4. Install Patrol CLI, which is needed to run Flutter integration tests that use Patrol's native UI automation feature. Use the latest prerelease version:

```
dart pub global activate patrol_cli
```

5. Remove the workaround for the this problem. To do this, open packages/patrol/lib/src/common.dart and remove [the following lines](https://github.com/leancodepl/patrol/blob/a49dfccb3b975c0d7f5f57e630f11b030a319b06/packages/patrol/lib/src/common.dart#L115-L121):

```
widgetTester.binding.platformDispatcher.onSemanticsEnabledChanged = () {
  // This callback is empty on purpose. It's a workaround for tests
  // failing on Android on Flutter 3.10 (#1265)
  //
  // See #1265 and #1352.
  // TODO(bartekpacia): Create bug report to Flutter.
};
```

6. `patrol test --target integration_test/example_test.dart` to run a simple integration test ([example_test.dart](https://github.com/leancodepl/patrol/blob/a49dfccb3b975c0d7f5f57e630f11b030a319b06/packages/patrol/example/integration_test/example_test.dart))
You might also want to replace the tests with the simple test that I have enclosed below.

7. Observe test failure in prompt and report respectively (see StackTraces below)

Related to: https://github.com/flutter/flutter/issues/129231

See also: https://github.com/leancodepl/patrol/issues/1474,  https://github.com/leancodepl/patrol/issues/1265 and https://github.com/flutter/flutter/issues/18015

### Expected results

Expects the test to pass and exit without any issues.

### Actual results

```console
No device specified, using the first one (emulator-5554)
• Building apk with entrypoint test_bundle.dart...
✓ Completed building apk with entrypoint test_bundle.dart (19.3s)
• Executing tests of apk with entrypoint test_bundle.dart on emulator-5554...
        See the report at: file:///.../build/app/reports/androidTests/connected/flavors/development/index.html
✗ Failed to execute tests of apk with entrypoint test_bundle.dart on emulator-5554 (Gradle test execution failed with code 1) (25.6s)
Exception: Gradle test execution failed with code 1
See the logs above to learn what happened. Also consider running with --verbose. If the logs still aren't useful, then it's a bug - please report it.
```

### Code sample

<details open><summary>Simple test sample</summary>

```dart
  group('Dummy tests =>', ()
  {
    patrolSetUp(() => ($) async {
      await $.native.initialize();
      await $.native.configure();
    });

    patrolWidgetTest(
      'find app text on widget',
      config: const PatrolTesterConfig(visibleTimeout: Duration(seconds: 10)),
          ($) async {
        await $.pumpWidgetAndSettle(
          MaterialApp(
            home: Scaffold(
              appBar: AppBar(title: const Text('app')),
              backgroundColor: Colors.blue,
            ),
          ),
        );
        expect($('app'), findsOneWidget);
        // TODO Fix
        // A SemanticsHandle was active at the end of the test.
        // All SemanticsHandle instances must be disposed by calling
        // dispose() on the SemanticsHandle.
      },
    );
  });
```

</details>

### Logs

<details open><summary>Logs</summary>

```console
java.lang.AssertionError: Dart test failed: app_test Dummy tests => find app text on widget
══╡ EXCEPTION CAUGHT BY FLUTTER TEST FRAMEWORK ╞═════════════════
The following assertion was thrown running a test:
A SemanticsHandle was active at the end of the test.
All SemanticsHandle instances must be disposed by calling
dispose() on the SemanticsHandle.

When the exception was thrown, this was the stack:
#0      WidgetTester._verifySemanticsHandlesWereDisposed (package:flutter_test/src/widget_tester.dart:1048:7)
#1      WidgetTester._endOfTestVerifications (package:flutter_test/src/widget_tester.dart:1039:5)
#2      TestWidgetsFlutterBinding._runTestBody (package:flutter_test/src/binding.dart:1025:22)
<asynchronous suspension>
#3      TestWidgetsFlutterBinding._createTestCompletionHandler.<anonymous closure> (package:flutter_test/src/binding.dart:804:12)
<asynchronous suspension>

The test description was:
find app text on widget
═════════════════════════════════════════════════════════════════
at pl.leancode.patrol.PatrolJUnitRunner.runDartTest(PatrolJUnitRunner.java:135)
at <applicationId>.MainActivityTest.runDartTest(MainActivityTest.java:29)
at java.lang.reflect.Method.invoke(Native Method)
at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
at org.junit.runners.Suite.runChild(Suite.java:128)
at org.junit.runners.Suite.runChild(Suite.java:27)
at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
at org.junit.runners.Suite.runChild(Suite.java:128)
at org.junit.runners.Suite.runChild(Suite.java:27)
at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
at org.junit.runner.JUnitCore.run(JUnitCore.java:115)
at androidx.test.internal.runner.TestExecutor.execute(TestExecutor.java:67)
at androidx.test.internal.runner.TestExecutor.execute(TestExecutor.java:58)
```

</details>

### Flutter Doctor output

<details open><summary>Doctor output</summary>

```console
Doctor summary (to see all details, run flutter doctor -v):
[√] Flutter (Channel stable, 3.16.3, on Microsoft Windows [Version 10.0.22631.3007], locale en-DK)
[√] Windows Version (Installed version of Windows is version 10 or higher)
[√] Android toolchain - develop for Android devices (Android SDK version 33.0.2)
[√] Chrome - develop for the web
[!] Visual Studio - develop Windows apps (Visual Studio Build Tools 2022 17.6.2)
    X Visual Studio is missing necessary components. Please re-run the Visual Studio installer for the "Desktop development with C++" workload, and include these components:
        MSVC v142 - VS 2019 C++ x64/x86 build tools
         - If there are multiple build tool versions available, install the latest
        C++ CMake tools for Windows
        Windows 10 SDK
[√] Android Studio (version 2022.2)
[√] IntelliJ IDEA Community Edition (version 2022.3)
[√] VS Code (version 1.77.3)
[√] Connected device (4 available)
[√] Network resources
```

</details>

### Patrol Doctor output

<details open><summary>Doctor output</summary>

```console
Patrol doctor:
Patrol CLI version: 2.6.2
Android:
• Program adb found in ...\AppData\Local\Android\Sdk\platform-tools\adb.exe
• Env var $ANDROID_HOME set to ...\AppData\Local\Android\Sdk
```

</details>

**Comment by huycozy** [comment]
Hi @nc-jeni 
When executing `flutter run`, I got this error:

```console
➜  example git:(master) ✗ flutter run 
Launching lib/main.dart on sdk gphone64 x86 64 in debug mode...
Running Gradle task 'assembleDebug'...                             21.6s
Exception: Gradle build failed to produce an .apk file. It's likely that this file was generated under
/Users/huynq/Documents/NEVERCODE.IO/RepoduceProjects/OPClone/patrol/packages/patrol/example/build, but the tool couldn't find it.
```

Do you have any idea to pass this error?

**Comment by nc-jeni** [comment]
> Hi @nc-jeni When executing `flutter run`, I got this error:
> 
> ```
> ➜  example git:(master) ✗ flutter run 
> Launching lib/main.dart on sdk gphone64 x86 64 in debug mode...
> Running Gradle task 'assembleDebug'...                             21.6s
> Exception: Gradle build failed to produce an .apk file. It's likely that this file was generated under
> /Users/huynq/Documents/NEVERCODE.IO/RepoduceProjects/OPClone/patrol/packages/patrol/example/build, but the tool couldn't find it.
> ```
> 
> Do you have any idea to pass this error?

Please have look at https://github.com/flutter/flutter/issues/129231

You might find the solution on that issue.

**Comment by bartekpacia** [comment]
Hi @nc-jeni, I see you're ﻿﻿using an old version of `patrol_cli`. Please check if this problem occurs on the latest ones.

**Comment by nc-jeni** [comment]
Hi @bartekpacia 
According to https://pub.dev/packages/patrol_cli/versions then v2.6.2 is the lastest?

**Comment by bartekpacia** [comment]
You have this in reproduction steps:

> Use the latest prerelease version:
>
> ```
> dart pub global activate \
>   --source git https://github.com/leancodepl/patrol.git \
>   --git-path packages/patrol_cli \
>   --git-ref patrol_cli-v2.0.0-dev.3
> ```

**Comment by nc-jeni** [comment]
> ```
> Patrol doctor:
> Patrol CLI version: 2.6.2
> Android:
> • Program adb found in ...\AppData\Local\Android\Sdk\platform-tools\adb.exe
> • Env var $ANDROID_HOME set to ...\AppData\Local\Android\Sdk
> ```

The reproduction step was copied from another issue, my mistake.
However, as stated in the bottom of the issue:

<details>
<summary>Patrol Doctor states</summary>
```
Patrol doctor:
Patrol CLI version: 2.6.2
Android:
• Program adb found in ...\AppData\Local\Android\Sdk\platform-tools\adb.exe
• Env var $ANDROID_HOME set to ...\AppData\Local\Android\Sdk
```
</details>

What version/branch did you fix it as I am still getting the issue on emulators? @bartekpacia

**Comment by bartekpacia** [comment]
It was many months ago and I really don't remember, apart from what I wrote in https://github.com/flutter/flutter/issues/129231 and https://github.com/leancodepl/patrol/issues/1265

**Comment by nc-jeni** [comment]
> It was many months ago and I really don't remember, apart from what I wrote in #129231 and [leancodepl/patrol#1265](https://github.com/leancodepl/patrol/issues/1265)

I guess, that the fix should be the versions I am using then.
Please re-visit it as it is obviously still an issue, @bartekpacia ?
Thanks

**Comment by bartekpacia** [comment]
I no longer work on Patrol.

**Comment by huycozy** [comment]
@nc-jeni I still can't pass the error above. I don't know which underlying issue is in the package. For the main issue, I would suggest you report it to package's maintainers first. If @bartekpacia's PR is not reverted in newer package versions, perhaps trying old Flutter versions is also a way to have more clues.

**Comment by nc-jeni** [comment]
> ```
> trol/example/b
> ```

@huycozy , instead of using the sample provided by Patrol then try to create your own integration test with the snippet I have provided in the description.

```
  group('Dummy tests =>', ()
  {
    patrolSetUp(() => ($) async {
      await $.native.initialize();
      await $.native.configure();
    });

    patrolWidgetTest(
      'find app text on widget',
      config: const PatrolTesterConfig(visibleTimeout: Duration(seconds: 10)),
          ($) async {
        await $.pumpWidgetAndSettle(
          MaterialApp(
            home: Scaffold(
              appBar: AppBar(title: const Text('app')),
              backgroundColor: Colors.blue,
            ),
          ),
        );
        expect($('app'), findsOneWidget);
        // TODO Fix
        // A SemanticsHandle was active at the end of the test.
        // All SemanticsHandle instances must be disposed by calling
        // dispose() on the SemanticsHandle.
      },
    );
```

**Comment by huycozy** [comment]
I ran your test but it succeed:

<details>
<summary>Demo (Pixel 4 API 34 emulator)</summary>


https://github.com/flutter/flutter/assets/104349824/c0d36094-e93c-4ec0-860e-b4370cc8e0b8


</details>

<details>
<summary>Output log</summary>

```console
➜  example git:(master) ✗ patrol test --target integration_test/simple_test.dart
No device specified, using the first one (emulator-5554)
• Building apk with entrypoint test_bundle.dart...
✓ Completed building apk with entrypoint test_bundle.dart (15.1s)
• Executing tests of apk with entrypoint test_bundle.dart on emulator-5554...
✓ Completed executing apk with entrypoint test_bundle.dart on emulator-5554 (119.0s)
➜  example git:(master) ✗ 
```
</details>

Patrol CLI version: 2.6.4

**Comment by darshankawar** [comment]
@nc-jeni 
Looking at above conversation, it looks like to be an issue with patrol framework and not flutter. I suggest you to report this in dedicated repo for further resolution. https://github.com/leancodepl/patrol/issues

If you are having an issue with flutter integration test framework, then we can look into the issue further. 
Closing from here for now, as it looks like third party package issue. If you disagree, write in comments and I'll reopen it.

**Comment by nc-jeni** [comment]
> > Patrol CLI version: 2.6.2
> 
> I try to bump to: Patrol CLI version: 2.6.4

Confirming that this is fixed by `Patrol CLI version: 2.6.4`
